### PR TITLE
[MRG] Nilearn masker cache bugfix

### PR DESCRIPTION
**Pull request by arthurmensch**
Should fix nilearn/nilearn#771

### CODE REVIEW

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 1]
```diff
@@ -315,3 +313,34 @@ def test_filter_and_mask():
...
+            masker.transform(this_niimg)
+            masker.transform(this_niimg)
+            masker.fit()
+            masker.transform(this_niimg)
+            l = os.listdir(join(temp_dir, 'joblib', 'nilearn',
+                                'input_data',
+                                'nifti_masker', 'filter_and_mask'))
+            # Ensure that filter and mask has not dumped twice
+            # the same data
+    assert_equal(len(l), 3)
```
Comment (arthurmensch) [inline nilearn/input_data/tests/test_nifti_masker.py]: I wanted to clear memory with a `mem.clear()` inside the for loop, but for some reason `func_code.py` does not appear in the second iteration if I do that -- which is bewildering.

File: nilearn/input_data/multi_nifti_masker.py [hunk 2]
```diff
@@ -211,7 +212,7 @@ def fit(self, imgs=None, y=None):
         else:
             self.affine_ = self.mask_img_.get_affine()
         # Load data in memory
-        self.mask_img_.get_data()
+        _safe_get_data(self.mask_img_)
```
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: This line does nothing.

File: nilearn/input_data/multi_nifti_masker.py [hunk 3]
```diff
@@ -201,7 +200,9 @@ def fit(self, imgs=None, y=None):
         # Resampling: allows the user to change the affine, the shape or both.
         if self.verbose > 0:
             print("[%s.transform] Resampling mask" % self.__class__.__name__)
-        self.mask_img_ = self._cache(image.resample_img)(
+        # self.mask_img_ is memapped initially :  we ensure it stays that way
+        # when using cache
+        self.mask_img_ = self._cache(image.resample_img, mmap_mode='r')(
```
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: I don't understand why you want the mask to be memmapped: it's not that big.

File: nilearn/image/resampling.py [hunk 4]
```diff
@@ -393,7 +395,7 @@ def resample_img(img, target_affine=None, target_shape=None,
     # We now know that some resampling must be done.
     # The value of "copy" is of no importance: output is always a separate
     # array.
-    data = img.get_data()
+    data = _safe_get_data(img)
```
Comment (AlexandreAbraham) [inline nilearn/image/resampling.py]: So you are saying that we should leave the choice to load data from disk or not to the user? Do we need to harmonize this behavior?

File: nilearn/input_data/nifti_masker.py [hunk 5]
```diff
@@ -247,7 +248,7 @@ def fit(self, imgs=None, y=None):
         else:
             self.affine_ = self.mask_img_.get_affine()
         # Load data in memory
-        self.mask_img_.get_data()
+        _safe_get_data(self.mask_img_)
```
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_masker.py]: Does nothing again.

File: nilearn/input_data/multi_nifti_masker.py (hunk 3 above)
Comment (arthurmensch) [inline nilearn/input_data/multi_nifti_masker.py]: Because `_utils.check_niimg_3d(self.mask_img) ` returns a memapped array, so self.mask_img_ is memmapped at first call and ndarary at second call if no parameter is passed

Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: The data array in the object is memmaped, which should be handled by joblib, right? I don't see the point in memmapping the Nifti1Image itself.

**Comment by AlexandreAbraham** [comment]
If I get it correctly:

  * you change the default behavior of always loading data by not loading data
  * you memmaped some cached results

I think that we are trying to fight a losing battle. Nifti1Image is not stateless and hashing them is not reliable. Data can be memmapped, or not. Headers can be different with no difference whatsoever on the data.

Small example:
```python
import joblib
import nibabel


def f(x):
    print('Not cached')


mem = joblib.Memory(cachedir='cache')
img1 = nibabel.load('file.nii')
img2 = nibabel.load('./file.nii')
mem.cache(f)(img1)
mem.cache(f)(img2)
```

Result:
```
Not cached
Not cached
```

My suggestion: writing a reliable hashing function for Nifti1Image once and for all. Plus, it can be clever by using filepaths when needed and by caching the data hash in the Nifti1Image header.

But I reckon by looking at the tests that this PR improves the default behavior.

File: nilearn/input_data/multi_nifti_masker.py [hunk 6]
```diff
@@ -201,6 +201,7 @@ def fit(self, imgs=None, y=None):
         # Resampling: allows the user to change the affine, the shape or both.
         if self.verbose > 0:
             print("[%s.transform] Resampling mask" % self.__class__.__name__)
+        print(hash(self.mask_img_))
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: You're forgetting to remove debug info :)

Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: Commits are labelled "WIP". I renamed the PR itself do that we don't review unfinished code.

File: nilearn/input_data/tests/test_multi_nifti_masker.py [hunk 7]
```diff
@@ -126,3 +127,61 @@ def test_joblib_cache():
         mask_hash = hash(masker.mask_img_)
         masker.mask_img_.get_data()
         assert_true(mask_hash == hash(masker.mask_img_))
+
+
+def test_transform_cache():
```
Comment (GaelVaroquaux) [inline nilearn/input_data/tests/test_multi_nifti_masker.py]: Also, I think that you could probably factor a bit this tests together: they contain a lot of duplicated code.

If you want you can create an auxiliary function that, given a masker and niimgs, fits and transforms thoses niimgs and raises an error if there is a recomputation.

**Comment by GaelVaroquaux** [comment]
@arthurmensch : you closed this PR, but don't forget to address our comments in the final PR :)

**Comment by arthurmensch** [comment]
Sorry I closed it by mistake

**Comment by AlexandreAbraham** [comment]
> But I think that it is a battle worth fighting.

Let me rephrase then: I don't think that trying to patch the code everywhere is the best solution. In 90% of the cases, we will use files from the hard-drive. In that case `hashing.hash(os.path.abspath(img.file_map['image'].filename))` is enough. If the image is not linked to a filename, then we can consider hashing the data.

> I also think that @arthurmensch is doing it right by writing a test for the desired functionality. This ensures that the functionality will stay.

Arthur is basically replacing the previous default behavior with a new one. I can also write use-cases that work with the previous system and not the new one: it's a matter of choice.

**Comment by arthurmensch** [comment]
I agree that using a custom hash for Nifti1Image would be a good idea, but the question becomes where to insert it : it is too specific for joblib, but I believe we don't want to insert some joblib-like code in nilearn

**Comment by arthurmensch** [comment]
Good news : it turns out that with no modification, with joblib bugfix, it seems to work. The issue was that I didn't properly monkey patch joblib in the test file (and still don't know how to do it)

File: nilearn/input_data/tests/test_multi_nifti_masker.py (hunk 7 above)
Comment (arthurmensch) [inline nilearn/input_data/tests/test_multi_nifti_masker.py]: Done
